### PR TITLE
ICU-22520 Move all localeID parsing logic into new ulocimp_getSubtags()

### PR DESCRIPTION
**Pull request by roubert**
The logic for parsing a localeID string into its constituent subtags is currently repeated over and over again in each one of the `uloc_get*()` functions, so that calling all these functions one after the other in order to get all the subtags does the parsing all over again from the beginning for each function call.

In order to avoid having to do this parsing over and over again, a lot of code instead has its own copy of the parsing logic in order to call the underlying `ulocimp_get*()` functions directly for lower runtime cost at the price of increased code complexity and repetition.

This new `ulocimp_getSubtags()` function, which writes natively to `icu::ByteSink` and has a convenience wrapper to write to `icu::CharString`, removes the repeated code from the `uloc_get*()` functions and makes it possible to update all code that calls the `ulocimp_get*()` functions.

##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22520
- [X] Required: The PR title must be prefixed with a JIRA Issue number.
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number.
- [X] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

ALLOW_MANY_COMMITS=true

### CODE REVIEW

**Comment by markusicu** [comment]
This sounds like a good idea for this code, operating on locale ID *strings*.
However, before I review it in detail, I wonder if this is what we want to or should do long-term.

It is good to consolidate subtags parsers, but effectively you are creating a full locale ID / language tag parser that does most of the same stuff as the C++ Locale constructor and .forLanguageTag().
Redundant code is not good. You are reducing it, but still.

So my question is, would we be better off making the C++ Locale constructor/factory parser good and efficient, and then change this sort-of-C-code to construct C++ Locale objects and use their normal functions to fetch the subtags?
This would let us focus on one implementation, and should make the calling code more similar to the Java code.

If we did that, we might need to add more StringPiece/string_view/ByteSink APIs to the C++ Locale. We could start by adding `@internal` functions; when we are happy with the new code, we could propose the new functions to be truly public. It might also give us another motivation to improve the structure and parser of class Locale (which we have wanted to do for a long time).

And maybe your new code could be repurposed as the new parser implementation in class Locale.

Adding @richgillam @macchiati @yumaoka @FrankYFTang to the discussion.

**Comment by roubert** [comment]
> However, before I review it in detail, I wonder if this is what we want to or should do long-term.

The way I see it, the changes in this PR are one further step towards making this code base possible to refactor, so it's a valuable change regardless of where we eventually want to end up in the future.

File: icu4c/source/common/uloc.cpp
Comment (markusicu) [inline icu4c/source/common/uloc.cpp:1198]: nit: Append(result.toStringPiece())  ?

File: icu4c/source/common/uloc.cpp
Comment (markusicu) [inline icu4c/source/common/uloc.cpp:1161]: optional: Consider *returning* the end pointer rather than writing it to an output argument.

File: icu4c/source/common/uloc.cpp
Comment (roubert) [reply icu4c/source/common/uloc.cpp:1198]: I'll soon replace it with writing directly to the sink instead.

File: icu4c/source/common/uloc.cpp
Comment (roubert) [reply icu4c/source/common/uloc.cpp:1161]: I'll keep that in mind, thanks!

File: icu4c/source/common/uloc.cpp
Comment (markusicu) [reply icu4c/source/common/uloc.cpp:1430]: On this item I do want to put my foot down. This is too obscure and surprising even for keeping it only temporarily. Please do add some kind of parameter for stopping after the region code / before the variants.

Comment (roubert) [reply icu4c/source/common/uloc.cpp:1430]: OK, I've now replaced the use of `ulocimp_forLanguageTag()` in `uloc_getVariant()` with a little more straightforward (and much less wasteful) code that instead of converting the entire language tag just checks for the `-t-`, `-u-` or `-x-` marker that indicates the start of a BCP-47 extension.

File: icu4c/source/common/uloc.cpp
Comment (markusicu) [inline icu4c/source/common/uloc.cpp:1109]: Should this be lenient with `-` vs `_`?
I *think* we usually are... but may be wrong.

File: icu4c/source/common/uloc.cpp
Comment (markusicu) [inline icu4c/source/common/uloc.cpp:1113]: ditto?

File: icu4c/source/common/uloc.cpp
Comment (markusicu) [inline icu4c/source/common/uloc.cpp:1559]: Why does the old and new code here bother to parse/identify the subtags at all, when eventually it just looks for `@` and parses the keywords from there?

File: icu4c/source/common/uloc.cpp
Comment (markusicu) [reply icu4c/source/common/uloc.cpp:1430]: I still don't like this asymmetric code. It looks like a bug, and someone else will stumble over it again.

Please check for pEnd here, like before, and if call sites want to stop between region & variant even though they want pEnd but not the variant, then please add a parameter for that.

**Review by markusicu — COMMENTED** [review]
praise: more good cleanup!

**Review by richgillam — APPROVED** [review]
I didn't give this quite as close a reading as maybe I should have, but since it seems like Markus has given it a very close reading, I don't feel so bad about that.

I really like what you're doing here; I think this simplifies and prettifies things quite a bit.

I wonder if we need a follow-on ticket.  It looks like you changed all the call sites that used the stuff in `ulocimp.h`, but I think there are some other places that use the public APIs that could also be simplified by using your new stuff in `ulocimp.h`.  (I think I may have written some of those.)  I think it might be worth it to check through the code for any such spots and either fix them or file tickets to go back and do that later.

**Comment by richgillam** [comment]
Actually, maybe I shouldn't have given an approval while you're still waiting for approval from Markus.  I think you should wait for him to sign off on this before checking in.

**Review by markusicu — APPROVED** [review]
thank you!!🎉

**Comment by roubert** [comment]
> It looks like you changed all the call sites that used the stuff in `ulocimp.h`, but I think there are some other places that use the public APIs that could also be simplified by using your new stuff in `ulocimp.h`.

That work has already been going on for a long time now and it is still going on, see eg. new PR #2812.

File: icu4c/source/common/uloc.cpp
Comment (roubert) [reply icu4c/source/common/uloc.cpp]: > I still think that it for the time being is a good idea to simply continue copy-pasting this recipe until all the locale memory management has been updated, for at that point in time we'll have a much clearer view of exactly what would be the most useful design for such a helper function.

PR #2863

**Pull request merged**
